cvat/apps/engine/serializers.py

```diff
@@ -667,16 +667,55 @@ def to_representation(self, instance):
             data['consensus_replicas'] = 0
 
         if request := self.context.get('request'):
-            perm = TaskPermission.create_scope_view(request, instance.segment.task)
-            result = perm.check_access()
-            if result.allow:
+            can_view_task = getattr(instance, "user_can_view_task", None)
+            if can_view_task is None:
+                perm = TaskPermission.create_scope_view(request, instance.segment.task)
+                result = perm.check_access()
+                can_view_task = result.allow
+
+            if can_view_task:
                 if task_source_storage := instance.get_source_storage():
                     data['source_storage'] = StorageSerializer(task_source_storage).data
                 if task_target_storage := instance.get_target_storage():
                     data['target_storage'] = StorageSerializer(task_target_storage).data
 
         return data
 
+    @classmethod
+    def many_init(cls, *args, **kwargs):
+        if len(args) == 1 and isinstance(args[0], list):
+            from .views import JobViewSet  # avoid circular import
+
+            ctx_request = kwargs.get("context", {}).get("request")
+            ctx_view = kwargs.get("context", {}).get("view")
+            if isinstance(ctx_view, JobViewSet) and ctx_view.action == "list" and ctx_request:
```

Comment:
Are the first two conditions here necessary? In what other circumstances would this method be called? Would it be inappropriate to use the same logic in them?

Reply:
I can say that Django does a 0 arg call during init, and `JobViewSet` can be uninitialized at that point. Maybe the checks can be simplified somehow, they just ensure we're in the supposed use case. We don't need them in 1-object calls, only in massive ones, like `list`.

Reply:
> We don't need this logic in 1-object calls, only in massive ones, like list.

This is the `many_init` method, which is only invoked when the serializer is created with `many=True`. So it seems like it should only get invoked in the scenario you want.

Also, I think the `len(args)` check could be avoided too by moving this logic to a more fitting place; you can give this serializer a custom `list_serializer_class`, and in that class, override `to_representation`.

Reply:
Ok, using `list_serializer_class` looks good to me.